Add v4 orders endpoint

joshuatf opened the pull request:
Fixes #950 

Adds the new endpoint so we can use new filters and return data from lookup tables.

### Screenshots
<img width="563" alt="screen shot 2019-01-09 at 2 32 29 pm" src="https://user-images.githubusercontent.com/10561050/50881066-728c9880-141b-11e9-891b-96af6018282a.png">

### Detailed test instructions:

1.  Make sure you have coupons and products associated with orders in your lookup tables.
2.  Rebuild your reports to ensure statuses are in the lookup tables.
3.  Send a request to `wp-json/wc/v4/reports/orders` and test the following params **(note the v4 namespace)**:
`before (date)`
`after (date)`
`page (int)`
`per_page (int)`
`orderby (date, num_items_sold, or net_total)`
`order (asc or desc)`
`status_is (string - status type without 'wc-' prefix)`
`status_is_not (string - status type without 'wc-' prefix)`
`product_includes (int or array)`
`product_excludes (int or array)`
`coupon_includes (int or array)`
`coupon_excludes (int or array)`
`customer_type (string 'returning' or 'new')`
`extended_info (bool)`
4.  Run phpunit tests to make sure everything is ✅

joshuatf committed:
Rename orders controllers and data store to order stats

joshuatf committed:
Add orders data store and controller

joshuatf committed:
Add coupon, product, and status filters

joshuatf committed:
Add customer type query filter

joshuatf committed:
Add extended info for products and categories

joshuatf committed:
Bump API namespace to v4

joshuatf committed:
Add missing schema properties

joshuatf committed:
Set default customer_type and extended_info params

joshuatf committed:
Add orders endpoint tests

timmyc commented on includes/data-stores/class-wc-admin-reports-orders-data-store.php:
```diff
@@ -233,320 +149,136 @@ public function get_data( $query_args ) {
-		if ( ! $order ) {
-			return;
+		if ( isset( $query_args['status_is_not'] ) && is_array( $query_args['status_is_not'] ) && count( $query_args['status_is_not'] ) > 0 ) {
+			$forbidden_statuses = array_map( array( $this, 'normalize_order_status' ), $query_args['status_is_not'] );
```
☠️ we do not speak of the forbidden_statuses

timmyc commented on includes/data-stores/class-wc-admin-reports-orders-data-store.php:
```diff
@@ -233,320 +149,136 @@ public function get_data( $query_args ) {
+			$extended_info               = new ArrayObject();
+			$extended_info['products']   = array();
+			$extended_info['categories'] = array();
+			foreach ( $order->get_items() as $item ) {
```
I think we had this discussion before on another PR ( with @psealock me thinks ), and I'm fine with doing this in a follow-up, but couldn't we be potentially duplicating some queries here for getting taxonomies?

timmyc replied:
The only idea I had for the categories again was to maybe pre-fetch all the categories prior to the foreach loops once, and create a map of them $terms[ id ] => $term->name sort of thing. Or perhaps even build up that mapping and one for the products as we loop through the orders.

Quite possibly I'm being overly paranoid here, and mysql or WP query caching will already help us out here.

joshuatf replied:
Took a stab at optimizing this a bit in https://github.com/woocommerce/wc-admin/pull/1246/commits/4c6119fd4f2d204acfb4fc6b3cc1b1223c5dd640 but it may need a bit more work.

Feedback is welcome 😄

timmyc reviewed (commented):
Quite the PR here Josh 😮 ! I've left a few comments that I figured I'd get posted while I do some manual testing against the endpoint. Also would love to get some more 👀 on this PR from perhaps @peterfabian and @claudiulodro

joshuatf committed:
Only use one JOIN if both excluded and included queries are used

joshuatf commented on includes/data-stores/class-wc-admin-reports-orders-data-store.php:
```diff
@@ -26,175 +20,101 @@ class WC_Admin_Reports_Orders_Data_Store extends WC_Admin_Reports_Data_Store imp
-				WHERE
-					{$wpdb->prefix}wc_order_coupon_lookup.coupon_id IN ({$excluded_coupons})
-				)";
+			$sql_query_params['from_clause']  .= " JOIN {$order_coupon_lookup_table} ON {$order_stats_lookup_table}.order_id = {$order_coupon_lookup_table}.order_id";
```
Smart thinking!  Updated this for coupons and products.

joshuatf committed:
Optimize query calls to get product and category data

claudiulodro reviewed (approved):
Works well in my testing and the data stores look good to me. 👍

joshuatf committed:
Split product and category fetch and limit to product_cat taxonomy

joshuatf commented:
Rebased and added https://github.com/woocommerce/wc-admin/pull/1246/commits/f7e08ba4f429072b050a963b6c23cfa0acf5ff0d which cleans up product/category mapping and fixes some mapping issues.

Could someone give this another look to make sure `extended_info` looks okay before we merge?

psealock commented:
@joshuatf I'm about halfway through review this, but need to run out to car mechanic. I'll pick this up again in a few hours

psealock reviewed (approved):
>Could someone give this another look to make sure extended_info is returning the right products and categories before we merge?

By looking at the code, the endpoint is returning whats I'm expecting. Nice work.
